File: examples/typescript/src/04-errors.ts
// Example 4: Functions with error throwing
// Tests shatter's ability to discover error paths and distinguish normal vs exceptional returns.

// EXPECTED BRANCHES:
//   1. denominator === 0                   → throws Error("division by zero")
//   2. !isFinite(numerator)                → throws Error("non-finite numerator")
//   3. result is integer                   → returns integer result
//   4. result is not integer               → returns float result

export function safeDivide(numerator: number, denominator: number): number {
    if (denominator === 0) {
        throw new Error("division by zero");
    }
    if (!isFinite(numerator)) {
        throw new Error("non-finite numerator");
    }
    const result = numerator / denominator;
    if (Number.isInteger(result)) {
        return result;
    }
    return result;
}

// EXPECTED BRANCHES:
//   1. items is empty                → throws Error("empty array")
//   2. any item < 0                  → throws Error("negative value")
//   3. all items === 0               → returns { sum: 0, avg: 0, max: 0 }
//   4. normal case, max > 100        → returns stats with "high-max" flag
//   5. normal case, max <= 100       → returns stats without flag

export function computeStats(items: number[]): {
    sum: number;
    avg: number;
    max: number;
    flag?: string;
} {
    if (items.length === 0) {
        throw new Error("empty array");
    }

    let sum = 0;
    let max = items[0];

    for (const item of items) {
        if (item < 0) {
            throw new Error("negative value");
        }
        sum += item;
        if (item > max) {
            max = item;
        }
    }

    const avg = sum / items.length;

    if (max === 0) {
        return { sum: 0, avg: 0, max: 0 };
    }

    if (max > 100) {
        return { sum, avg, max, flag: "high-max" };
    }

    return { sum, avg, max };
}

File: examples/standalone/rust/04_errors.rs
// Example 4: Functions with error handling via Result.
// Tests shatter's ability to discover error paths and distinguish normal vs exceptional returns.

/// safe_divide — 4 branches: denominator==0→error, numerator not finite→error,
/// integer result→Ok, non-integer result→Ok.
fn safe_divide(numerator: f64, denominator: f64) -> Result<f64, String> {
    if denominator == 0.0 {
        return Err("division by zero".to_string());
    }
    if numerator.is_infinite() || numerator.is_nan() {
        return Err("non-finite numerator".to_string());
    }
    Ok(numerator / denominator)
}

struct Stats {
    sum: f64,
    avg: f64,
    max: f64,
    flag: Option<String>,
}

/// compute_stats — 5 branches: empty slice→error, any negative→error,
/// all zeros→{0,0,0}, max>100→stats with "high-max" flag, else→stats.
fn compute_stats(items: &[f64]) -> Result<Stats, String> {
    if items.is_empty() {
        return Err("empty array".to_string());
    }

    let mut sum = 0.0;
    let mut max = items[0];

    for &item in items {
        if item < 0.0 {
            return Err("negative value".to_string());
        }
        sum += item;
        if item > max {
            max = item;
        }
    }

    let avg = sum / items.len() as f64;

    if max == 0.0 {
        return Ok(Stats { sum: 0.0, avg: 0.0, max: 0.0, flag: None });
    }

    if max > 100.0 {
        return Ok(Stats { sum, avg, max, flag: Some("high-max".to_string()) });
    }

    Ok(Stats { sum, avg, max, flag: None })
}

fn main() {
    println!("{:?}", safe_divide(10.0, 3.0));
    println!("{:?}", safe_divide(10.0, 0.0));
    match compute_stats(&[1.0, 2.0, 150.0]) {
        Ok(s) => println!("sum={}, avg={}, max={}, flag={:?}", s.sum, s.avg, s.max, s.flag),
        Err(e) => println!("Error: {e}"),
    }
}
